WIP: Add kolektor dataset

Ravindu987 opened the pull request:
# Description

- Work in progress
- Added the dataset classes, dataloaders and other necessary functions to load `Kolektor Dataset`
- Due to the nature of the dataset, used unconventional methods which needs to be discussed
- Requires ImageIO and Numpy libraries

- Fixes #964 

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [ ] My code follows the [pre-commit style and check guidelines](https://github.com/openvinotoolkit/anomalib/blob/main/CONTRIBUTING.md) of this project.
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes
- [ ] I have added a summary of my changes to the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) (not for minor changes, docs and tests).

ashwinvaidya17 commented on src/anomalib/data/kolektor.py:
```diff
@@ -0,0 +1,179 @@
+from __future__ import annotations
```
The file should include the source to the original dataset and a license header. You can have a look here https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/data/mvtec.py

ashwinvaidya17 commented on src/anomalib/data/kolektor.py:
```diff
@@ -0,0 +1,179 @@
+            transform=transform_eval,
+            split=Split.TEST,
+            root=root,
+        )
```
This is missing `prepare_data` method. Have a look at https://github.com/openvinotoolkit/anomalib/blob/27adbaeb7da406c52bceba06dbc875ba3c0e955a/src/anomalib/data/mvtec.py#L258 to see an example implementation

ashwinvaidya17 commented on src/anomalib/data/kolektor.py:
```diff
@@ -0,0 +1,179 @@
+        samples.loc[samples.label_index == 1]
+        .apply(lambda x: Path(x.image_path).stem in Path(x.mask_path).stem, axis=1)
+        .all()
+    ), "Mismatch between anomalous images and ground truth masks. Make sure the mask files in 'ground_truth' \
+              folder follow the same naming convention as the anomalous images in the dataset (e.g. image: '000.png', \
+              mask: '000.png' or '000_mask.png')."
```
Are you sure this message is correct?

ashwinvaidya17 reviewed (changes requested):
Thanks for the efforts! I understand this is WIP, I've still gone ahead and added a few comments. Also, can you also accress the issues raised by Codacy.

Ravindu987 commented:
Hi @ashwinvaidya17 
Thank you very much for your feedback.
In the latest commits I resolved three of the issues you raised.

Regarding the `prepare_dataset` there's an issue. The download link for the dataset doesn't provide an extension for the downloaded file. Instead it sends the extension via the response header. This works well with a browser but the `urlretrieve` method used in the `download_and_extract` method downloads the file without extension and therefore cannot extract.

We have two options to resolve this
1. Instead of `download_and_extract` use a custom function to download and set the extension.
2. Update the `download_and_extract` function to receive the response header and set the extension. If we choose this we'll have to update other `make_dataset` methods accordingly.

Please let me know your thoughts on this.

Also as I have informed @samet-akcay , I had to pause the work on this until I get a green light on the current logical implementation ( As you might have seen, I had to classify images on the runtime instead of loading from directories because of the nature of the dataset). Please let me know whether the logic is good to go so that I can finish up.

Ravindu987 commented:
Hey guys
I've gone ahead and completed writing docstrings and implemented a custom function to download and extract the dataset.
Since I've tested it locally I removed the WIP tag. Please have a look.

Could you refer me a documentation on the test convention used here. Once I get an insight on this I can finish it as well.

Thanks

samet-akcay commented on src/anomalib/data/kolektor.py:
```diff
@@ -0,0 +1,315 @@
+    return 0
+
+
+def download_and_extract_kolektor(url: str, root: Path, file_hash: str):
+    """Download and extract a dataset.
+
+    Args:
+        url  (Str): Url for directly downloading the dataset
+        root (Path): Root directory where the dataset will be stored.
+        file_hash (Str): MD5 Checksum of the correct zip file
+    """
+    root.mkdir(parents=True, exist_ok=True)
+
+    downloaded_file_path = Path.joinpath(root, "KolektorSDD.zip")
+    if downloaded_file_path.exists():
+        logger.info("Existing dataset archive found. Skipping download stage.")
+    else:
+        logger.info("Downloading the Kolektor dataset archive")
+        urlretrieve(  # nosec - suppress bandit warning (urls are hardcoded)
+            url=url,
+            filename=downloaded_file_path,
+        )
+        logger.info("Checking the hash of the downloaded file.")
+        hash_check(downloaded_file_path, file_hash)
+
+    logger.info("Extracting dataset into given folder")
+    with ZipFile(downloaded_file_path, "r") as zip_file:
+        zip_file.extractall(root)
+
+    logger.info("Cleaning up files.")
+    (downloaded_file_path).unlink()
```
There is a function to download and extract data [here](https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/data/utils/download.py#L220). Is there a specific difference between your implementation and this implementation? If not, this would be a duplication.

Ravindu987 replied:
There's a slight difference as I've mentioned [here](https://github.com/openvinotoolkit/anomalib/pull/983#issuecomment-1498479796) The `download_and_extract` function in `utils` downloads the Kolektor dataset without an extension and therefore it can't be extracted. This happens since the provided link for the dataset doesn't specify the extension which is required by the `download_and_extract` function. In this implementation I've hardcoded the .zip extension in line 65.

samet-akcay replied:
Maybe we could update the existing code a little bit. How about the following:

Replace
https://github.com/openvinotoolkit/anomalib/blob/d81dbb5b3dfd831576d137fb93881b34254a7fff/src/anomalib/data/utils/download.py#L25-L31

with:
```python
@dataclass
class DownloadInfo:
    """Info needed to download a dataset from a url."""

    name: str
    url: str
    hash: str
    filename: str | None = None
```

And replace
https://github.com/openvinotoolkit/anomalib/blob/d81dbb5b3dfd831576d137fb93881b34254a7fff/src/anomalib/data/utils/download.py#L229-L230

with,
```python
    # save the compressed file in the specified root directory, using the same file name as on the server
    if info.filename:
        downloaded_file_path = root / info.filename
    else:
        downloaded_file_path = root / info.url.split("/")[-1]
```

Then, I think `anomalib`'s download function should work. You wouldn't need to overwrite it.

Let me know if it works. Cheers!

Ravindu987 replied:
Thank you for the great suggestion. I've implemented and tested the change and it works fine. It has been updated in the latest commits.

samet-akcay commented on src/anomalib/data/kolektor.py:
```diff
@@ -0,0 +1,315 @@
+        if (self.root).is_dir():
+            logger.info("Found the dataset.")
+        else:
+            download_and_extract_kolektor(
+                "https://go.vicos.si/kolektorsdd",
+                self.root,
+                "2b094030343c1cd59df02203ac6c57a0",
+            )
```
If you define `DOWNLOAD_INFO` variable as given below

```python
DOWNLOAD_INFO = DownloadInfo(
    name="kolektor",
    url="https://go.vicos.si/kolektorsdd",
    hash="2b094030343c1cd59df02203ac6c57a0",
)
```

Then it would be possible to use `anomalib.data.download.download_and_extract` function as:

```python
download_and_extract(self.root, DOWNLOAD_INFO)
```

This would be inline with the rest of the datamodule implementations. For reference, here are the some of the implementations:
[mvtec-ad](https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/data/mvtec.py#L52), [avenue](https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/data/avenue.py#L44), [btech](https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/data/btech.py#L38), [mvtec-3d](https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/data/mvtec_3d.py#L49)...

Ravindu987 replied:
This links with the previous issue which makes the `download_and_extract` function not usable here.

samet-akcay reviewed (changes requested):
Thanks @Ravindu987 for your great contribution! I have a comment regarding the use of download and extract function.

samet-akcay commented on src/anomalib/data/kolektor.py:
```diff
@@ -0,0 +1,315 @@
+
+
+# Check if a mask shows defects
+def is_good(path):
```
Not sure if this name is descriptive enough. Maybe your comment could be used as the name of the function such as `is_mask_anomalous or something similar?

samet-akcay commented on src/anomalib/data/kolektor.py:
```diff
@@ -0,0 +1,315 @@
+
+# Check if a mask shows defects
+def is_good(path):
+    img_arr = imread(path)
```
is there any specific reason for using `imageio.imread` function? Is this in anomalib's list of requirements?

Ravindu987 replied:
Actually there was no specific reasons. Pillow or opencv seemed overkill for simply loading the image. Please let me know if I should use any other library.

Ravindu987 replied:
I didn't thought it would make much of a time difference here. In that case it would be good to switch. I'll do a small time comparison and switch to opencv if there's a significant difference.

Ravindu987 replied:
Opencv gave around a 25% decrease in time usage so I switched to opencv in the latest commit.

Ravindu987 commented:
Thank you very much @samet-akcay  for the review. I resolved one requested change and answered your questions as well. Please let me know your thoughts and any further changes.

samet-akcay reviewed (approved):
lgtm, thanks a lot for your contribution!

Ravindu987 commented:
Thanks for the review guys. 
The Codacy Static Code Analysis fails because  the `init` function in the DataModule class has too many arguments. I see no way around this as other implementations have the same issue.
Not sure why the Tox checks are failing. When I ran all pre-merge checks locally they all passed. Any ide on this?

samet-akcay commented:
> Thanks for the review guys. The Codacy Static Code Analysis fails because the `init` function in the DataModule class has too many arguments. I see no way around this as other implementations have the same issue. Not sure why the Tox checks are failing. When I ran all pre-merge checks locally they all passed. Any ide on this?

It's a problem with our CI that we need to fix. You don't need to do anything else at this point. Thank again for your contribution

Ravindu987 commented:
> > Thanks for the review guys. The Codacy Static Code Analysis fails because the `init` function in the DataModule class has too many arguments. I see no way around this as other implementations have the same issue. Not sure why the Tox checks are failing. When I ran all pre-merge checks locally they all passed. Any ide on this?
> 
> It's a problem with our CI that we need to fix. You don't need to do anything else at this point. Thank again for your contribution

Noted with thanks. Hope to contribute more in the future.
